File: android/app/src/main/java/dev/ghostty/connect/terminal/TmuxPassthroughParser.kt
package dev.ghostty.connect.terminal

import java.io.ByteArrayOutputStream

internal class TmuxPassthroughParser(private val onBytes: (ByteArray) -> Unit) {
    private val plain = ByteArrayOutputStream()
    private val passthrough = ByteArrayOutputStream()
    private var prefixIndex = 0
    private var capturing = false
    private var pendingEscape = false
    private var draining = false

    fun feed(bytes: ByteArray) {
        bytes.forEach(::accept)
        flushPlain()
    }

    fun reset() {
        plain.reset()
        passthrough.reset()
        prefixIndex = 0
        capturing = false
        pendingEscape = false
        draining = false
    }

    private fun accept(value: Byte) {
        if (capturing) {
            if (pendingEscape) {
                pendingEscape = false
                when (value) {
                    ESC -> append(ESC) // tmux escapes payload ESC bytes by doubling them.
                    ST_END -> finishPassthrough()
                    else -> {
                        append(ESC)
                        append(value)
                    }
                }
            } else if (value == ESC) {
                pendingEscape = true
            } else {
                append(value)
            }
            return
        }

        if (value == PREFIX[prefixIndex]) {
            prefixIndex++
            if (prefixIndex == PREFIX.size) {
                flushPlain()
                prefixIndex = 0
                capturing = true
                passthrough.reset()
            }
            return
        }
        if (prefixIndex > 0) {
            plain.write(PREFIX, 0, prefixIndex)
            prefixIndex = 0
            if (value == PREFIX[0]) {
                prefixIndex = 1
                return
            }
        }
        plain.write(value.toInt())
    }

    private fun append(value: Byte) {
        if (draining) return
        if (passthrough.size() >= MAX_PASSTHROUGH_BYTES) {
            draining = true
            passthrough.reset()
        } else {
            passthrough.write(value.toInt())
        }
    }

    private fun finishPassthrough() {
        if (!draining) onBytes(passthrough.toByteArray())
        passthrough.reset()
        capturing = false
        draining = false
    }

    private fun flushPlain() {
        if (plain.size() == 0) return
        onBytes(plain.toByteArray())
        plain.reset()
    }

    companion object {
        private val PREFIX = "\u001bPtmux;".toByteArray()
        private const val MAX_PASSTHROUGH_BYTES = 7 * 1024 * 1024
        private const val ESC: Byte = 0x1b
        private const val ST_END: Byte = 0x5c
    }
}

File: ios/GhosttyConnect/Terminal/TmuxPassthroughParser.swift
import Foundation

final class TmuxPassthroughParser {
    private static let prefix = Array("\u{1b}Ptmux;".utf8)
    private static let maxBytes = 7 * 1_024 * 1_024
    private let onBytes: (Data) -> Void
    private var plain: [UInt8] = []
    private var passthrough: [UInt8] = []
    private var prefixIndex = 0
    private var capturing = false
    private var pendingEscape = false
    private var draining = false

    init(onBytes: @escaping (Data) -> Void) { self.onBytes = onBytes }

    func feed(_ data: Data) {
        data.forEach(accept)
        flushPlain()
    }

    private func accept(_ byte: UInt8) {
        if capturing {
            if pendingEscape {
                pendingEscape = false
                if byte == 0x1b { append(0x1b) }
                else if byte == 0x5c { finish() }
                else { append(0x1b); append(byte) }
            } else if byte == 0x1b { pendingEscape = true }
            else { append(byte) }
            return
        }
        if byte == Self.prefix[prefixIndex] {
            prefixIndex += 1
            if prefixIndex == Self.prefix.count {
                flushPlain()
                prefixIndex = 0
                capturing = true
                passthrough.removeAll(keepingCapacity: true)
            }
            return
        }
        if prefixIndex > 0 {
            plain.append(contentsOf: Self.prefix.prefix(prefixIndex))
            prefixIndex = 0
            if byte == Self.prefix[0] { prefixIndex = 1; return }
        }
        plain.append(byte)
    }

    private func append(_ byte: UInt8) {
        guard !draining else { return }
        guard passthrough.count < Self.maxBytes else { passthrough.removeAll(); draining = true; return }
        passthrough.append(byte)
    }

    private func finish() {
        if !draining { onBytes(Data(passthrough)) }
        passthrough.removeAll(keepingCapacity: true)
        capturing = false
        draining = false
    }

    private func flushPlain() {
        guard !plain.isEmpty else { return }
        onBytes(Data(plain))
        plain.removeAll(keepingCapacity: true)
    }
}
